### PR TITLE
fix: Multiple Leaks around FairGeoLoader

### PR DESCRIPTION
**Pull request by ChristianTackeGSI**
Also some small other stylistic changes

---

Checklist:

* [X] Followed the [Contributing Guidelines](https://github.com/FairRootGroup/FairRoot/blob/dev/CONTRIBUTING.md)

### CODE REVIEW

File: fairroot/geobase/FairGeoLoader.h [hunk 1]
```diff
@@ -16,34 +16,40 @@
...
 class FairGeoLoader : public TNamed
 {
   public:
     FairGeoLoader(const char* Name, const char* title);
     FairGeoLoader();
     ~FairGeoLoader() override;
-    FairGeoInterface* getGeoInterface() { return fInterface; }
-    FairGeoBuilder* getGeoBuilder() { return fGeoBuilder; }
+    FairGeoInterface* getGeoInterface() { return &fInterface; }
+    FairGeoBuilder* getGeoBuilder() { return fGeoBuilder.get(); }
```
[the remark this answers is not in the document]
Comment (ChristianTackeGSI) [reply fairroot/geobase/FairGeoLoader.h:40]: This is a public API, we can't easily change the return type. And really a raw pointer is a good fit for an optional reference.

File: fairroot/geobase/FairGeoLoader.cxx [hunk 2]
```diff
@@ -31,40 +31,33 @@ FairGeoLoader* FairGeoLoader::Instance() { return fgInstance; }
...
+        fGeoBuilder = std::make_unique<FairGeoRootBuilder>("TGeo builder", "geometry builder", std::move(geom));
     } else if (strncmp(Name, "G3Native", 8) == 0) {
         cout << "-I- FairGeoLoader() : Native G3 Geometry is used: This option is not supported any more!" << endl;
         exit(0);
         //        gGeoManager = NULL;
         //        fGeoBuilder=new FairGeoG3Builder("G3 builder","geometry builder");
     }
 
-    fInterface->setGeomBuilder(fGeoBuilder);
+    fInterface.setGeomBuilder(fGeoBuilder.get());
```
[the remark this answers is not in the document]
Comment (ChristianTackeGSI) [reply fairroot/geobase/FairGeoLoader.cxx:56]: Take a look at the member variables as described in FairGeoLoader.h! `fInterface` and `fGeoBuilder` are defined there, and their order is changed there.

File: fairroot/geobase/FairGeoLoader.cxx [hunk 3]
```diff
@@ -31,40 +31,33 @@ FairGeoLoader* FairGeoLoader::Instance() { return fgInstance; }
...
-        fGeoBuilder = new FairGeoRootBuilder("TGeo builder", "geometry builder");
-        (static_cast<FairGeoRootBuilder*>(fGeoBuilder))->setGeoManager(geom);
+        auto geom = std::make_unique<TGeoManager>("FAIRGeom", "FAIR geometry");
+        fGeoBuilder = std::make_unique<FairGeoRootBuilder>("TGeo builder", "geometry builder", std::move(geom));
     } else if (strncmp(Name, "G3Native", 8) == 0) {
         cout << "-I- FairGeoLoader() : Native G3 Geometry is used: This option is not supported any more!" << endl;
         exit(0);
         //        gGeoManager = NULL;
         //        fGeoBuilder=new FairGeoG3Builder("G3 builder","geometry builder");
     }
```
Comment (dennisklein) [inline fairroot/geobase/FairGeoLoader.cxx:54]: I feel this should go into a helper, e.g.:

```cpp
static std::unique_ptr<FairGeoBuilder> FairGeoLoader::MakeGeoBuilder(const char* name)
{
    if (strncmp(Name, "TGeo", 4) == 0) {
        auto geom = std::make_unique<TGeoManager>("FAIRGeom", "FAIR geometry");
        return std::make_unique<FairGeoRootBuilder>("TGeo builder", "geometry builder", std::move(geom));
    }

    if (strncmp(Name, "G3Native", 8) == 0) {
        cout << "-I- FairGeoLoader() : Native G3 Geometry is used: This option is not supported any more!" << endl;
        exit(0);
        //        gGeoManager = NULL;
        //        fGeoBuilder=new FairGeoG3Builder("G3 builder","geometry builder");
    }

    return {};
}
```

What do you think?

File: fairroot/geobase/FairGeoRootBuilder.h [hunk 4]
```diff
@@ -10,30 +10,29 @@
...
-    TGeoManager* geoManager;   // ROOT geometry manager
+    std::unique_ptr<TGeoManager> geoManager;   //!< ROOT geometry manager
 
   public:
     FairGeoRootBuilder();
     FairGeoRootBuilder(const char*, const char*);
+    FairGeoRootBuilder(const char*, const char*, std::unique_ptr<TGeoManager>);
     ~FairGeoRootBuilder() override = default;
-    void setGeoManager(TGeoManager* me) { geoManager = me; }
+    [[deprecated]] void setGeoManager(TGeoManager* me) { geoManager.reset(me); }
```
Comment (dennisklein) [inline fairroot/geobase/FairGeoRootBuilder.h:35]: Should this deprecation be added to the editorial changelog?

Comment (ChristianTackeGSI) [reply fairroot/geobase/FairGeoRootBuilder.h:35]: I really hope that this code is actually only "internal FairRoot code". So I wanted to keep the amount of "paper work" to a minimum (I was actually thinking of just removing it at all).

**Review by karabowi — DISMISSED** [review]

**Comment by ChristianTackeGSI** [comment]
Do we want to backport this to 19.0? If no, I would postpone merging this until the release is done.

File: fairroot/geobase/FairGeoLoader.cxx (hunk 3 above)
Comment (dennisklein) [reply fairroot/geobase/FairGeoLoader.cxx:54]: My point is about double initialization of the unique_ptr member.

**Review by dennisklein — APPROVED** [review]

**Comment by ChristianTackeGSI** [comment]
Still the question, whether we should backport (fast-forward) this in 19.0?

**Comment by dennisklein** [comment]
What would speak against?

**Comment by ChristianTackeGSI** [comment]
I don't know anything. If it's a good enough fix for dev, then it should be good enough for 19.0 also.

Maybe because 19.0 is in some sort of "freeze"?
But still it's a bug fix.

**Pull request merged**